### DISCUSSION TITLE
Path to beatmap grouping in song select

### DISCUSSION THREAD

**Discussion by LittleEndu** [ppy/osu#14713] [Q&A]
Right now to implement beatmap grouping in song select I see two major blockers. One is quite obvious - there's no (final) design for it yet. The other not so obvious problem is that `BeatmapCarousel`/`CarouselRoot` since its creation in 2016/2018 has only dealt with beatmap sets. Carousel doesn't directly deal with `CarouselItem`s so there's no way to implement new behaviour around something else being in `CarouselRoot`.

Here's a list of things, in no particular order, that would need to be changed to implement grouping:
 * `loadBeatmapSets()` can't add only `CarouselBeatmapSet`s into `CarouselRoot`
 * `DrawableCarouselBeatmap` along with any other new type of carousel item should reside directly in `CarouselScrollContainer` instead of its parent
   * There should be a pool for any subclass of `DrawableCarouselItem` instead of only for sets
   * All items should implement `UpdateItem()` / No item should implement `CreateDrawableRepresentation()` 
 * `updateYPositions()` should be recursive over any type of carousel item in the root
 
Any of these would be a huge change to the way carousel works and can't be implemented in complete isolation. The way forward would probably be to: remove `DrawableCarouselBeatmap` from `DrawableCarouselBeatmapSet` -> rewrite `updateYPositions()` -> get all visible items regardless of their type -> get pooled drawable for that item and add it to scroll container.

Earlier today I thought that at least I can rework `updateYPositions` as I have dealt with that in the past, but I didn't know the logic to `Update()` the Y position is in `DrawableCarouselBeatmapSet` and not in `DrawableCarouselItem` (which after allowing any type of item to be in the scroll container is probably where it should reside)

**Comment by LittleEndu** [comment]
Another can of worms is selection although `SelectNext` shouldn't be that difficult to reimplement once `CarouselRoot` can have any type of `CarouselItem` in it. But there's weird edge cases where two difficulties of same set can be next to each other so should pressing <kbd>Right</kbd> go to next set or next difficulty:
![screenshot223](https://user-images.githubusercontent.com/11320483/132898131-1e82d875-6c8f-4f79-9cbd-103c40fd0fec.png)

**Comment by bdach** [comment]
Theorycrafting how this should work has little benefit if the design isn't final IMO. For one, I don't know that the carousel should try to hold everything at once. Because [if this draft is to be taken at face value](https://www.figma.com/file/DXKwqZhD5yyb1igc3mKo1P/Song-Select-Draft-2?node-id=0%3A1), then difficulties change appearance *entirely* when ungrouped compared to when they're grouped under their respective sets. So how is that transition going to play out? Is each difficulty going to have two different visual representations?

I'd honestly consider keeping the current view as-is and then adding an alternative "grouped" carousel, and then switching between them.

**Reply by LittleEndu** [reply]
I don't think it matters where the non-consecutive display logic exists, I just thought sharing as much of the logic between grouped and no grouping cases made more sense. 

But if a new carousel is to be written, most of the same problems still exists on top of now having to split `BeatmapCarousel` into three (parent class and two subclasses for grouped or not)

**Reply by peppy** [reply]
I not sure this is something you can take on. If/when it is to be done, it will be someone with enough knowledge and insight to implement it.

As I said in discord, I can't answer any questions without doing it myself. Even the structural changes you say in this thread don't necessarily sound right.

**Reply by LittleEndu** [reply]
> Even the structural changes you say in this thread don't necessarily sound right.

I was thinking a bit too general when I initially thought how this could be implemented. But bdach mentioning, that there's no way to transition from grouped drawable design to no grouping or vice-versa, and that there could really just be two carousels made me realize that carousel root only needs to hold either beatmap sets (which hold difficulties) or a new type of groups (which also would hold difficulties just in their non-consecutive design/model). Other functions just need to consider the the group of difficulties might not be a beatmap set.

> I not sure this is something you can take on. If/when it is to be done, it will be someone with enough knowledge and insight to implement it.

What insight other than final design am I missing that I couldn't take this on??? This attitude is really confusing me right now...

**Reply by peppy** [reply]
Let me try again: As per what I said on discord, I would need to implement the system to see how this works and how it should work. I don't think it's productive to discuss how the system should be implemented or structured as such. You should just give it a go if you think you can make it work well.

What i'm getting at is that this discussion thread is not necessary.

The idea of giving structural advice when neither the design or intended UX has been properly understood is a waste of time. In addition, I have a realm branch which considerably changes the code surrounding the model fetch / load portion of the existing carousel, so I'm not sure I can recommend refactoring it directly at the current point in time.

If you are looking to figure a good UX, consider making a proof-of-concept implementation in an isolated test scene using empty boxes. That would be the starting point for me.

In addition, flyte's designs have zero animation, zero UX flow, zero nothing. They aren't designs - they are mock-up screenshots with most of the important things left to guess-work in terms of UX.

**Comment by LittleEndu** [comment]
Crazy thing that would work is to separate all sets with however many beatmaps they have into sets with just one beatmap: https://github.com/ppy/osu/compare/master...LittleEndu:carousel-grouping

![osu_2021-09-11_18-01-26](https://user-images.githubusercontent.com/11320483/132952166-b82565e0-5a16-476d-a0ad-81d18e631a6e.png)

This is good enough for what I want to use grouping for (I want non-consecutive sorting) so this is where I'll stop. I don't think something this silly will get merged so there's no reason I should implement actual "grouping under the same header" part on that branch.
